subprojects/core/src/integTest/groovy/org/gradle/api/provider/PropertyStateIntegrationTest.groovy

```diff
@@ -0,0 +1,372 @@
+                text = '$OUTPUT_FILE_CONTENT'
+                outputFiles = files("\$buildDir/helloWorld.txt")
+            }
+
+            task consumer(type: Consumer) {
+                inputFiles = producer.outputs.files
+            }
+
+            class Producer extends DefaultTask {
+                @Input
+                String text
+
+                private final PropertyState<FileCollection> outputFiles = project.property(FileCollection)
+                
+                void setOutputFiles(FileCollection outputFiles) {
+                    this.outputFiles.set(outputFiles)
+                }
+                
+                @OutputFiles
+                FileCollection getOutputFiles() {
+                    outputFiles.get()
+                }
+
+                @TaskAction
+                void produce() {
+                    getOutputFiles().each {
+                        it << text
+                    }
+                }
+            }
+            
+            class Consumer extends DefaultTask {
+                @InputFiles
+                FileCollection inputFiles
+                
+                @TaskAction
+                void consume() {
+                    inputFiles.each {
+                        println it.text
+                    }
+                }
+            }
+        """
+
+        when:
+        succeeds('consumer')
+
+        then:
+        executedTasks.containsAll(':producer', ':consumer')
+        outputContains(OUTPUT_FILE_CONTENT)
+    }
+
+    static String customGroovyBasedTaskType() {
+        """
+            import org.gradle.api.DefaultTask
+            import org.gradle.api.file.FileCollection
+            import org.gradle.api.provider.PropertyState
+            import org.gradle.api.tasks.TaskAction
+            import org.gradle.api.tasks.Input
+            import org.gradle.api.tasks.OutputFiles
+
+            class MyTask extends DefaultTask {
+                private PropertyState<Boolean> enabled = project.property(Boolean)
+                private PropertyState<FileCollection> outputFiles = project.property(FileCollection)
+                
+                MyTask() {
+                    enabled.set(false)
+                    outputFiles.set(project.files())
+                }
+
+                @Input
+                boolean getEnabled() {
+                    enabled.get()
+                }
+                
+                void setEnabled(PropertyState<Boolean> enabled) {
+                    this.enabled = enabled
+                }
+                
+                void setEnabled(boolean enabled) {
+                    this.enabled.set(enabled)
+                }
+                
+                @OutputFiles
+                FileCollection getOutputFiles() {
+                    outputFiles.get()
+                }
+
+                void setOutputFiles(PropertyState<FileCollection> outputFiles) {
+                    this.outputFiles = outputFiles
```

Comment:
not really required (`FileCollection` is-a `Provider<Set<File>>` which is covered below). Could instead be:

```
void setOutputFiles(Iterable<File> files) { this.outputFiles.set(files); }
```